git: add deprecation warnings for `git.subprocess`

emilazy opened the pull request:
~~Only intended for merge when we decide to start the deprecation process, but~~ should be ready ~~for code review now.~~

~~The first commit could be split out into another PR if we want to merge it now.~~
<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the commits rather than adding commits on top. Use force-push when
pushing the updated commits (`jj git push` does that automatically when you
rewrite commits). Merge the PR at will once it's been approved. See
https://github.com/jj-vcs/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

emilazy commented:
Undrafting this since @thoughtpolice expressed interest. This is of course not intended for merge before the upcoming patch release.

I put 0.30 as the version based on the Discord discussion, but that should just be treated as a placeholder for whatever’s decided.

PhilipMetzger commented:
I think the outlined four releases are good enough for me. Since the last time I expressed stronger opinions on it.

emilazy commented:
Rebased for conflicts, and added some docs updates. This should be ready now, if we want to do it. https://github.com/jj-vcs/jj/pull/6048 shows the rest of the owl.

yuja commented on cli/src/git_util.rs:
```diff
@@ -275,6 +275,24 @@ pub fn with_remote_git_callbacks<T>(ui: &Ui, f: impl FnOnce(git::RemoteCallbacks
+            });
+        };
+        callbacks.git2_deprecation_warning = Some(&mut git2_deprecation_warning);
+    }
```
I don't think we need callback. Can't we instead add `warn(ui, &git_settings)` function and let `cmd_git_*()` functions call it?

Maybe we can also use `debug.suppress-git2-deprecation-warning` config key instead (because `UserSettings` should be available there.)

emilazy replied:
Yeah, that works, and produces slightly nicer output. Done. (This also means that `with_git_subprocess` could have been a regular method after all, but it doesn’t feel worth the churn to change it.)

yuja commented on cli/src/cli_util.rs:
```diff
@@ -2866,6 +2866,27 @@ pub fn print_unmatched_explicit_paths<'a>(
+        )?;
+    }
+    Ok(())
+}
```
nit: might be simpler to move this function to `git_util`, and handle `cfg(feature = "git2")` internally, but I don't have strong feeling.

emilazy replied:
I somewhat like the `#[cfg]` because it makes it easier to see what’s due for removal. `git_util` doesn’t depend on `cli_util` (which contains `WorkspaceCommandHelper`) currently (but `cli_util` does depend on `git_util`), do you think breaking that property to do the move would be okay?

yuja replied:
I think `settings` can be passed in to this function instead. I tend to avoid inserting `cfg(..)` at random places, but that's just my preference.

emilazy replied:
I moved it and adjusted the signature. I agree about `#[cfg]` in general, but I think in this case it helps make removal a little more mechanical, so I left it for now.

yuja reviewed (approved):
Thanks.

emilazy commented:
Actually, if we’re doing a point release for the gitoxide issue, should we wait to land this until after that’s done?
